### ISSUE TITLE
diffsol JAX wrapper

### ISSUE THREAD

**Issue by martinjrobins** [martinjrobins/diffsol#303]
child issue of #131 for the JAX wrapper. See under development [here](https://github.com/jackbmontgomery/diffsol-jax)

**Comment by martinjrobins** [comment]
Hi  @jackbmontgomery, just creating this issue to we don't pollute the parent issue too much. I wanted to let you know that the diffsol-c adjoint features are in place for you to start using this, which should simplify the issues you've been having:
- The methods that you need are [solve_adjoint_fwd](https://docs.rs/diffsol-c/latest/diffsol_c/ode/struct.OdeWrapper.html#method.solve_adjoint_fwd) and [solve_adjoint_bkwd](https://docs.rs/diffsol-c/latest/diffsol_c/ode/struct.OdeWrapper.html#method.solve_adjoint_bkwd)
- `solve_adjoint_fwd` produces an [AdjointCheckpoint](https://docs.rs/diffsol-c/latest/diffsol_c/adjoint_checkpoint/struct.AdjointCheckpointWrapper.html) that is consumed by solve_adjoint_bkwd
- These methods are on the main [OdeWrapper](https://docs.rs/diffsol-c/latest/diffsol_c/ode/struct.OdeWrapper.html) struct which actually jit compiles the diffsl code. This is serializable as well to a snapshot, so you can store it as raw bytes if you want (if needed). Let me know if you need anything else serializable as well (e.g. AdjointCheckpoint)
- diffsol-c uses [HostArray](https://docs.rs/diffsol-c/latest/diffsol_c/host_array/struct.HostArray.html) for vectors/matrices. See [pydiffsol](https://github.com/alexallmont/pydiffsol) for examples of how to convert these to numpy arrays

**Comment by jackbmontgomery** [comment]
Hi @martinjrobins, I was working on the wrapper this weekend and I've run into some issues with the adjoint
API that I'd like to discuss some more.

The wrapper calls `solve_adjoint_fwd` on the forward pass and `solve_adjoint_bkwd` on the backward
pass. The forward returns the `Checkpointing` object, which lives on the Rust heap, and I pass a raw
pointer to it through JAX's computation graph as an integer, then hand it back to the backward call.

JAX/XLA doesn't give the guarantees a raw pointer needs. It can duplicate, drop, reorder, or replay
calls. That means the checkpoint leaks under forward-mode autodiff, and can be freed twice (or used
after free) on the backward.

I think the ideal pattern is for the backward pass to be a pure function of a residual value that it
threads from forward to backward itself:

    fwd(params, t_span) -> (ys, residual)
    bwd(residual, g_ys)  -> grad_params

You mention that the checkpoints can be serialised, I think that this would would be the solution to
the problem. The one constraint from the JAX side is that a residual buffer must have a size known
at trace time. I can see two issues with this in the current diffsol-c api:

1. From what I can see in `method.rs`, checkpoints are pushed off the solver's internal adaptive
   steps, not the requested output times. So the checkpoint count grows with however many internal
   steps which is only known when the solver finishes.

2. The segment in Checkpointing fields hold a HermiteInterpolator, which stores the (ys, ydots, ts)
   for a checkpoint interval and that's a variable number of internal steps, so it's another
   variable-sized object. My understanding is that these segments are a recomputed cache rather than
   essential state: on the backward pass, interpolate() re-solves between the bracketing checkpoints
   to rebuild a segment. If that's right, then they wouldn't need to be serialized at all and a
   reconstructed Checkpointing could start with empty segments and rebuild them as the backward
   sweep needs. That would remove the variable-sized part, leaving just checkpoints: Vec<State> +
   terminal_reset_root_idx to serialize.

A fixed-size serialisation of the checkpoints seems like it would make the adjoint usable from JAX.
For it to have a static size I think it needs two things:

1. To skip the Hermite segments and recompute them on the backward and
2. A predictable checkpoint count. Two directions I can think of are:

- a. the caller asks for a fixed N checkpoints, with diffsol maintaining N slots as steps accumulate
  size is exactly N states or
- b. keep the current scheme but add a max_checkpoints cap and pad/record the actual count

**Comment by martinjrobins** [comment]
The most common use case for the checkpointing is to have two checkpoints, one at the start and one at the end, so restricting it to two would work in the majority of cases. The main issue is skipping the hermite segments, as this defeats the purpose of using checkpointing in the first place. You might as well just not store anything if you are going to run the forward solve again during the backwards pass.

Can you point me towards any documentation / code /examples that details the reqreuiments for what the forward pass can store?

Happy to have a chat on Teams if its useful. Alex (from pydiffsol) and I have a regular meeting on Wednesday, I can forward you the link

**Comment by martinjrobins** [comment]
one possible workaround for the residual static size restriction, could you store the checkpointing from diffsol in a global cache with a unique key, then pass that key as the residual? the backward pass then gets the key and extracts the checkpointing info from the cache

**Comment by jackbmontgomery** [comment]
Hi @martinjrobins, I am thinking about how I could drop the LLVM dependency for the JAX wrapper.

As I understand it, LLVM is pulled in only for Enzyme. Cranelift already emits rhs_grad, so the forward solve works on it and the adjoint computation is the only thing forcing LLVM, since rhs_rgrad / rhs_srgrad are Enzyme-only.

But I already trace f(t, u, p) in JAX, so I can get those functions from JAX's function transformations which I can then lower to jaxpr and convert to diffsl like I do for the rhs function already. We can then jit compile these functions with cranelift. This would then remove the LLVM dependency from the package.

diffsol-c does not have an OdeWrapper constructor accepts injected derivatives.  Would a creation route that takes the rhs plus these derivatives be feasible and do you think this is a good idea for the package?

**Comment by martinjrobins** [comment]
its a good idea, but the diffsl language has no way of saying "this tensor is actually the gradient of F wrt the parameters" or similar.

however, you *can* supply all your raw functions (rhs, rhs grad, rhs sens etc) as a static or dynamic library, so if you are able to go from jaxpr to this (using cranelift or more likely whatever jit method that JAX uses) then you can use the existing diffsol-c API.

**Comment by martinjrobins** [comment]
one thing you might want to consider is bypassing diffsl entirely. You can get StableHLO from jax easy enough:

```python
 From Python: Generate HLO from your JAX function:
lowered = jax.jit(f).lower(args)
hlo = lowered.hlo()                                   # XlaComputation
stablehlo_bytes = lowered.stablehlo().as_serialized()  # StableHLO MLIR bytecode
```

So you just need a way to JIT StableHLO, JAX uses XLA so you might as well do the same. A proper example is here (https://github.com/openxla/xla/blob/main/xla/examples/axpy/stablehlo_compile_test.cc), summarised by an LLM is:

```c++
std::unique_ptr<xla::PjRtClient> client = xla::GetXlaPjrtCpuClient(options);
mlir::OwningOpRef<mlir::ModuleOp> program = /* deserialize bytecode here */;
std::unique_ptr<xla::PjRtLoadedExecutable> executable =
    client->CompileAndLoad(xla::MaybeOwningMlirModule(std::move(program)),
                           xla::CompileOptions{}).value();
auto out = executable->Execute({{arg0.get(), arg1.get()}}, /*options=*/{}).value();
```

**Comment by martinjrobins** [comment]
doing all this from rust is probably a pain, but I can see some bindings already exist (in an early state): https://github.com/eaplatanios/ryft

**Comment by martinjrobins** [comment]
You could add a struct in diffsol that implements all the right `OdeEquation` traits, and has a constructor with the serialized stablehlo bytecode. Then you just need a constructor for `OdeWrapper` in diffsol-c

**Comment by jackbmontgomery** [comment]
Hi @martinjrobins, sorry for the silence over the past few weeks. I have a working on a potential route here that would remove the LLVM dependency and just wanted to share it with you before going further.

I didn't go via XLA/PjRt in the end, since XLA's CPU backend is itself built on LLVM and does not ship with jax. Instead I've been testing [elodin's cranelift-mlir](https://github.com/elodin-sys/elodin/tree/main/libs/cranelift-mlir): a rust StableHLO to Cranelift JIT compiler. Elodin built it to run JAX-defined physics models natively inside their simulator, which is more or less exactly our problem shape.

The idealised flow for diffsol-jax would be:
  1. The user writes `f(t, u, p)` in ordinary JAX.
  2. The wrapper derives everything diffsol needs using JAX's own transformations, the jvp for the forward/implicit path and the vjps with respect to `u` and `p` for the adjoint.
  3. Each function is lowered to StableHLO text via `jax.jit(f).lower(...)`, and JIT-compiled with cranelift-mlir.
  4. The resulting native function pointers are handed to diffsol through the external symbol interface, and `solve` / `solve_adjoint_fwd` / `solve_adjoint_bkwd`.

On a small MLP RHS this works end to end: the forward RHS and the full vjp both match JAX.

There are some shortcoming that I should note given it would be a very important dependency.
- **Correctness bugs exist in cranelift-mlir's reverse-mode surface.**  There were some issues with the dimensions when lowering the vjp. This is not really of note for Elodin because they do not use the gradient features of jax but just have forward only simulations. They are small bugs but just makes me think of the maintenance issues for this dependency.
- **It's a git dependency.** The crate is `publish = false` inside elodin's monorepo
- **Its StableHLO parser is hand-rolled** over the text format, so drift in what JAX emits between versions is a risk. Naturally, we can just pin specific versions but worth noting
- **CPU only.** Cranelift emits CPU code, so this route never gives GPU execution.

What do you think of this? Do you think this is worth migrating the library to? Do you think we could make a struct that implements the
OdeEquation traits that can use this?

**Comment by martinjrobins** [comment]
hmmm, its interesting, but I don't think its a good long term solution, one of the main selling points of JAX is that you can scale up to GPU easily so a CPU-only JAX wrapper would have limited applicability, and I want to built up the GPU features of diffsol in the coming years. And yes I would worry about the fact that its a hand-rolled StableHLO parser that would be prone to correctness errors. You would have users writing arbitrary JAX code for the ODE equations, so you would want it to match JAX rigorously. 

It is certainly a pain that XLA is built on LLVM in terms of wheel size, but the size should be comparable to jaxlib (which bundles XLA/LLVM), so it should be fine. And then you would have some surity that you are matching JAX exactly.

what do you mean that jax does not ship with XLA's cpu backend? Surely jaxlib does though? What does it use for cpu jit then?

**Comment by martinjrobins** [comment]
looks like jaxlib includes both xla and llvm from their build file: https://github.com/jax-ml/jax/blob/main/jaxlib/BUILD

**Comment by jackbmontgomery** [comment]
Yes sorry, I did not phrase this correctly. You're right that jaxlib bundles XLA/LLVM for CPU jit. What I meant is that it isn't exposed in a form we can use from outside: the XLA CPU client is statically linked into jaxlib's python extension, which only exports python entry points. I was hoping jax would ship a standalone cpu plugin the way they do for [GPU](https://pypi.org/project/jax-cuda12-pjrt/).

But yes, I completely understand the focus on GPU and batching. I think the only way to guarantee correctness for an arbitrary jax rhs is to execute the exact StableHLO jax produces on the same runtime, like you suggested originally.

I have this working as a prototype now:
1. The rust side loads a PJRT plugin at runtime via the [pjrt crate](https://docs.rs/pjrt/latest/pjrt/). The CPU plugin binary comes prebuilt from gopjrt releases just so I don't need to build it myself.
2. The python side traces the rhs with jax's own transformations into the module family diffsol needs, serializes each with jax.export, and hands the bytes to rust.
3. Rust compiles them on the plugin and exposes them through a struct implementing the OdeEquations trait family.

Module outputs matches jax and the adjoint matches finite differences, including for an MLP rhs.

I think this would also scale to GPU because the PJRT C API is backend-agnostic because the same code path can load the CUDA plugin.

Also just to note, this is not on diffsol-jax right now. It's a large change so I have kept it in a separate repo while testing the pattern and since it needs a different ode wrapper constructor in diffsol-c.

**Comment by martinjrobins** [comment]
great work @jackbmontgomery, this workflow sounds perfect!

**Comment by jackbmontgomery** [comment]
I have pushed this change to [diffsol-pjrt](https://github.com/jackbmontgomery/diffsol-pjrt). I will eventually put it into diffsol-jax but because I am now using diffsol not diffsol-c as the backend I am leaving it. When there is a constructor for the ode wrapper then I'll do the rewrite in diffsol-jax. Is that fine with you @martinjrobins?

**Comment by martinjrobins** [comment]
Yup, that sounds great. Would the order be:
1. Add your `StableHloEquations` struct to diffsol (similar to the existing `DiffSl` struct)
2. Add a constructor to the diffsl-c `OdeWrapper` that takes in modules and creates `StableHloEquations` internally

**Comment by jackbmontgomery** [comment]
Yes that makes sense. I am happy to work on these as well, if you are happy with that?
